diff --git a/benchmarks/linear_programming/cuopt/run_mip.cpp b/benchmarks/linear_programming/cuopt/run_mip.cpp
index 8e217f275..73cbe951b 100644
--- a/benchmarks/linear_programming/cuopt/run_mip.cpp
+++ b/benchmarks/linear_programming/cuopt/run_mip.cpp
@@ -92,17 +92,18 @@ void write_to_output_file(const std::string& out_dir,
 
 inline auto make_async() { return std::make_shared<rmm::mr::cuda_async_memory_resource>(); }
 
-void read_single_solution_from_path(const std::string& path,
-                                    const std::vector<std::string>& var_names,
-                                    std::vector<std::vector<double>>& solutions)
+// reads a solution from an input file. The input file needs to be csv formatted
+// var_name,val
+std::vector<double> read_solution_from_file(const std::string file_path,
+                                            const std::vector<std::string>& var_names)
 {
   solution_reader_t reader;
-  bool success = reader.read_from_sol(path);
+  bool success = reader.readFromCsv(file_path);
   if (!success) {
-    CUOPT_LOG_ERROR("Initial solution reading error!");
+    CUOPT_LOG_INFO("Initial solution reading error!");
+    exit(-1);
   } else {
-    CUOPT_LOG_INFO(
-      "Success reading file %s Number of var vals %lu", path.c_str(), reader.data_map.size());
+    CUOPT_LOG_INFO("Success reading csv! Number of var vals %lu", reader.data_map.size());
   }
   std::vector<double> assignment;
   for (auto name : var_names) {
@@ -111,42 +112,20 @@ void read_single_solution_from_path(const std::string& path,
     if ((it != reader.data_map.end())) {
       val = it->second;
     } else {
-      CUOPT_LOG_TRACE("Variable %s has no input value ", name.c_str());
+      CUOPT_LOG_INFO("Variable %s has no input value ", name.c_str());
       val = 0.;
     }
     assignment.push_back(val);
   }
-  if (assignment.size() > 0) {
-    CUOPT_LOG_INFO("Adding a solution with size %lu ", assignment.size());
-    solutions.push_back(assignment);
-  }
-}
-
-// reads a solution from an input file. The input file needs to be csv formatted
-// var_name,val
-std::vector<std::vector<double>> read_solution_from_dir(const std::string file_path,
-                                                        const std::string& mps_file_name,
-                                                        const std::vector<std::string>& var_names)
-{
-  std::vector<std::vector<double>> initial_solutions;
-  std::string mps_file_name_no_ext = mps_file_name.substr(0, mps_file_name.find_last_of("."));
-  // check if a directory with the given mps file exists
-  std::string initial_solution_dir = file_path + "/" + mps_file_name_no_ext;
-  if (std::filesystem::exists(initial_solution_dir)) {
-    for (const auto& entry : std::filesystem::directory_iterator(initial_solution_dir)) {
-      read_single_solution_from_path(entry.path(), var_names, initial_solutions);
-    }
-  } else {
-    read_single_solution_from_path(file_path, var_names, initial_solutions);
-  }
-  return initial_solutions;
+  CUOPT_LOG_INFO("Adding a solution with size %lu ", assignment.size());
+  return assignment;
 }
 
 int run_single_file(std::string file_path,
                     int device,
                     int batch_id,
                     std::string out_dir,
-                    std::optional<std::string> initial_solution_dir,
+                    std::optional<std::string> input_file_dir,
                     bool heuristics_only,
                     int num_cpu_threads,
                     bool write_log_file,
@@ -184,35 +163,23 @@ int run_single_file(std::string file_path,
     CUOPT_LOG_ERROR("Parsing MPS failed exiting!");
     return -1;
   }
-  if (initial_solution_dir.has_value()) {
-    auto initial_solutions = read_solution_from_dir(
-      initial_solution_dir.value(), base_filename, mps_data_model.get_variable_names());
-    for (auto& initial_solution : initial_solutions) {
-      bool feasible_variables =
-        test_constraint_and_variable_sanity(mps_data_model,
-                                            initial_solution,
-                                            settings.get_absolute_tolerance(),
-                                            settings.get_relative_tolerance(),
-                                            settings.get_integrality_tolerance());
-      if (feasible_variables) {
-        settings.add_initial_solution(initial_solution.data(), initial_solution.size());
-      }
-    }
+  if (input_file_dir.has_value()) {
+    auto initial_solution =
+      read_solution_from_file(input_file_dir.value(), mps_data_model.get_variable_names());
+    settings.set_initial_solution(initial_solution.data(), initial_solution.size());
+    test_constraint_sanity(mps_data_model,
+                           initial_solution,
+                           settings.get_absolute_tolerance(),
+                           settings.get_relative_tolerance(),
+                           settings.get_integrality_tolerance());
   }
 
   settings.set_time_limit(time_limit);
   settings.set_heuristics_only(heuristics_only);
   settings.set_num_cpu_threads(num_cpu_threads);
   settings.set_log_to_console(log_to_console);
-  cuopt::linear_programming::benchmark_info_t benchmark_info;
-  settings.set_benchmark_info_ptr(&benchmark_info);
   auto start_run_solver = std::chrono::high_resolution_clock::now();
   auto solution         = cuopt::linear_programming::solve_mip(&handle_, mps_data_model, settings);
-  CUOPT_LOG_INFO(
-    "first obj: %f last improvement of best feasible: %f last improvement after recombination: %f",
-    benchmark_info.objective_of_initial_population,
-    benchmark_info.last_improvement_of_best_feasible,
-    benchmark_info.last_improvement_after_recombination);
   // solution.write_to_sol_file(base_filename + ".sol", handle_.get_stream());
   std::chrono::milliseconds duration;
   auto end = std::chrono::high_resolution_clock::now();
@@ -232,9 +199,7 @@ int run_single_file(std::string file_path,
   std::stringstream ss;
   int decimal_places = 2;
   ss << std::fixed << std::setprecision(decimal_places) << base_filename << "," << sol_found << ","
-     << obj_val << "," << benchmark_info.objective_of_initial_population << ","
-     << benchmark_info.last_improvement_of_best_feasible << ","
-     << benchmark_info.last_improvement_after_recombination << "\n";
+     << obj_val << "\n";
   write_to_output_file(out_dir, base_filename, device, batch_id, ss.str());
   CUOPT_LOG_INFO("Results written to the file %s", base_filename.c_str());
   return sol_found;
@@ -244,7 +209,7 @@ void run_single_file_mp(std::string file_path,
                         int device,
                         int batch_id,
                         std::string out_dir,
-                        std::optional<std::string> initial_solution_dir,
+                        std::optional<std::string> input_file_dir,
                         bool heuristics_only,
                         int num_cpu_threads,
                         bool write_log_file,
@@ -258,7 +223,7 @@ void run_single_file_mp(std::string file_path,
                                   device,
                                   batch_id,
                                   out_dir,
-                                  initial_solution_dir,
+                                  input_file_dir,
                                   heuristics_only,
                                   num_cpu_threads,
                                   write_log_file,
@@ -313,11 +278,7 @@ int main(int argc, char* argv[])
     .scan<'i', int>()
     .default_value(-1);
 
-  program.add_argument("--initial-solution-path")
-    .help(
-      "Path to the initial solution file expressed in .sol format. This will also check a "
-      "directory named the same as the input mps file (without the .mps extension) and add all "
-      "solutions in that directory to the initial solution pool.");
+  program.add_argument("--initial-solution-path").help("path to the initial solution csv file");
 
   program.add_argument("--heuristics-only")
     .help("run heuristics only (t/f)")
@@ -334,7 +295,7 @@ int main(int argc, char* argv[])
 
   program.add_argument("--log-to-console")
     .help("log to console (t/f)")
-    .default_value(std::string("t"));
+    .default_value(std::string("f"));
 
   program.add_argument("--time-limit")
     .help("time limit")
@@ -367,7 +328,6 @@ int main(int argc, char* argv[])
   int num_cpu_threads  = program.get<int>("--num-cpu-threads");
   bool write_log_file  = program.get<std::string>("--write-log-file")[0] == 't';
   bool log_to_console  = program.get<std::string>("--log-to-console")[0] == 't';
-  if (write_log_file) { log_to_console = false; }
 
   if (program.is_used("--out-dir")) {
     out_dir     = program.get<std::string>("--out-dir");
@@ -380,9 +340,9 @@ int main(int argc, char* argv[])
   }
 
   int n_batches = program.get<int>("--n-batches");
-  std::optional<std::string> initial_solution_dir;
+  std::optional<std::string> initial_solution_file;
   if (program.is_used("--initial-solution-path")) {
-    initial_solution_dir = program.get<std::string>("--initial-solution-path");
+    initial_solution_file = program.get<std::string>("--initial-solution-path");
   }
 
   if (run_dir) {
@@ -448,7 +408,7 @@ int main(int argc, char* argv[])
                                gpu_id,
                                batch_num,
                                out_dir,
-                               initial_solution_dir,
+                               initial_solution_file,
                                heuristics_only,
                                num_cpu_threads,
                                write_log_file,
@@ -477,7 +437,7 @@ int main(int argc, char* argv[])
                     0,
                     0,
                     out_dir,
-                    initial_solution_dir,
+                    initial_solution_file,
                     heuristics_only,
                     num_cpu_threads,
                     write_log_file,
